services/venice-router/src/main/java/com/linkedin/venice/router/MetaDataHandler.java

```diff
@@ -477,6 +491,66 @@ private void handleCurrentVersionLookup(ChannelHandlerContext ctx, VenicePathPar
     setupResponseAndFlush(OK, OBJECT_MAPPER.writeValueAsBytes(response), true, ctx);
   }
 
+  /**
+   * Handles the discovery of blob transfer nodes based on store settings.
+   * Retrieves host names for live DVC nodes ready to transfer blobs.
+   * Only returns host names from nodes that have completed a full push and are active.
+   *
+   * @return a response with a list of host names for live DVC nodes; returns an empty list if no live nodes are found or if conditions are not met
+   */
+  private void handleBlobDiscovery(ChannelHandlerContext ctx, VenicePathParserHelper helper, HttpRequest request)
+      throws IOException {
+
+    // i.e. /blob_discovery?store_name=storeName&version=22&partition=2
+    Map<String, String> queryParams = helper.extractQueryParameters(request);
+    String storeName = queryParams.get(NAME);
+    String storeVersion = queryParams.get(VERSION);
+    String storePartition = queryParams.get(STORE_PARTITION);
+
+    if (StringUtils.isEmpty(storeName) || StringUtils.isEmpty(storeVersion) || StringUtils.isEmpty(storePartition)) {
+      byte[] errBody = ("Blob Discovery: missing storeName, storeVersion, or storePartition " + storeName + storeVersion
+          + storePartition).getBytes();
+      setupResponseAndFlush(BAD_REQUEST, errBody, false, ctx);
+      return;
+    }
+
+    Store store = storeRepository.getStore(storeName);
+    if (store == null || !store.isBlobTransferEnabled() || store.isHybrid()) {
+      byte[] errBody =
+          ("Blob Discovery: store: " + storeName + " could not be found in cluster: " + clusterName).getBytes();
+      setupResponseAndFlush(NOT_FOUND, errBody, false, ctx);
+      return;
+    }
+
+    if (!store.isBlobTransferEnabled() || store.isHybrid()) {
+      byte[] errBody =
+          ("Blob Discovery: blob transfer is not enabled or store: " + storeName + " is not batch-only store")
+              .getBytes();
+      setupResponseAndFlush(FORBIDDEN, errBody, false, ctx);
+      return;
+    }
+
+    BlobDiscoveryResponse response = new BlobDiscoveryResponse();
+    try {
+      // gets the instances for a FULL_PUSH for the store's version and partitionId
+      // gets the instance's hostnames from its keys & filter to include only live instances
+      Map<CharSequence, Integer> instances = pushStatusStoreReader
+          .getPartitionStatus(storeName, Integer.parseInt(storeVersion), Integer.parseInt(storePartition), null);
+      List<String> liveNodeHostNames = instances.entrySet()
+          .stream()
+          .map(Map.Entry::getKey)
+          .map(CharSequence::toString)
+          .filter(instanceHostName -> pushStatusStoreReader.isInstanceAlive(storeName, instanceHostName))
+          .collect(Collectors.toList());
+      response.setLiveNodeNames(liveNodeHostNames);
+    } catch (VeniceException e) {
+      response.setError(
```

Comment:
we don't have error in `BlobDiscoveryResponse`. Can you add it?

Reply:
`BlobDiscoveryResponse` is extending from [ControllerResponse](https://jarvis.corp.linkedin.com/codesearch/result/?name=ControllerResponse.java&path=venice%2Finternal%2Fvenice-client-common%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fvenice%2Fcontrollerapi&reponame=linkedin%2Fvenice#64) which has the `setError()`

All of the "Responses" under `controllerapi` path extend from ControllerResponse. I.E. [MultiSchedmaIdResponse](https://jarvis.corp.linkedin.com/codesearch/result/?name=MultiSchemaIdResponse.java&path=venice%2Finternal%2Fvenice-client-common%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fvenice%2Fcontrollerapi&reponame=linkedin%2Fvenice) & [SystemStoreHeartbeatResponse](https://jarvis.corp.linkedin.com/codesearch/result/?name=SystemStoreHeartbeatResponse.java&path=venice%2Finternal%2Fvenice-client-common%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fvenice%2Fcontrollerapi&reponame=linkedin%2Fvenice)